pyfr/backends/cuda/cublaslt.py

```diff
@@ -0,0 +1,296 @@
+            raise ValueError('Matrices too large for cuBLAS')
+
+        # α and β factors for D = α*(A*B) + β*C
+        if a.dtype == np.float64:
+            alpha_ct, beta_ct = c_double(alpha), c_double(beta)
+            ctype = w.CUBLAS_COMPUTE_64F
+            dtype = w.CUDA_R_64F
+        else:
+            alpha_ct, beta_ct = c_float(alpha), c_float(beta)
+            ctype = w.CUBLAS_COMPUTE_32F
+            dtype = w.CUDA_R_32F
+
+        desc = GEMMDesc(None, None, None, None, None, None, 0)
+
+        # Convenience wrapper
+        def gemm(stream):
+            w.cublasLtMatmul(
+                h, desc.mm, byref(alpha_ct), A, desc.a, B,
+                desc.b, byref(beta_ct), C, desc.c, C, desc.c,
+                byref(desc.algo), desc.ws_ptr, desc.ws_size, stream
+            )
+
+        # Obtain the performance of the kernel
+        try:
+            desc, dt = self._mul_cache[ckey]
+        except KeyError:
+            # Create matrix layouts
+            a_desc = CUBLASLtMatrixLayout(self, h, A, dtype)
+            b_desc = CUBLASLtMatrixLayout(self, h, B, dtype)
+            c_desc = CUBLASLtMatrixLayout(self, h, C, dtype)
+
+            # Matmul descriptor
+            mm_desc = CUBLASLtMatmulDesc(self, h, ctype, dtype)
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSA] = w.CUBLAS_OP_N
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSB] = w.CUBLAS_OP_N
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSC] = w.CUBLAS_OP_N
+
+            # Heuristic preference descriptor
+            pref = CUBLASLtMatmulPreference(self, h)
+            pref[w.CUBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES] = 1024 * 1024 * 4
+
+            # Get matmul heuristics
+            heurs = (CUBLASLtMatmulHeuristicResult * self.nkerns)()
+            nreturn = c_int()
+            w.cublasLtMatmulAlgoGetHeuristic(
+                h, mm_desc, a_desc, b_desc, c_desc, c_desc, pref, self.nkerns,
+                heurs, byref(nreturn)
+            )
+
+            # Save a copy of the contents of the output matrix
+            out_np = getattr(out, 'parent', out).get()
+
+            # Benchmark the kernel
+            best_kern = None
+            for heur in heurs[:nreturn.value]:
+                # Allocate gemm workspace
+                ws_size = heur.workspaceSize
+                ws = CUDADevAlloc(cuda, ws_size) if ws_size else None
+
+                desc = GEMMDesc(
+                    mm_desc, a_desc, b_desc, c_desc, heur.algo, ws, ws_size
+                )
+
+                dt = self._benchmark(gemm)
+                if best_kern is None or dt < best_kern[-1]:
+                    best_kern = desc, dt
+
+            # Delete the potentially large hueristics array
+            del heurs
+
+            # Restore the output matrix
+            getattr(out, 'parent', out).set(out_np)
+
+            # Update the cache
+            desc, dt = best_kern
+            self._mul_cache[ckey] = best_kern
+
+        class MulKernel(CUDAKernel):
```

Comment:
Need to allocate workspace (the exact amount) around here and then pass it to `MulKernel` to prevent memory leaks.  Current approach has race conditions if two kernels (such as two M0 instances) both execute at the same time (which will happen when we do gradient interpolation).  Given the sizes both will hit in the cache and so they will have the same workspace.